### PR TITLE
fix: added protocols to the url along with correct port range

### PR DESCRIPTION
**Pull request by allroundexperts**
### Fixed Issues
$ https://github.com/Expensify/App/issues/17205

# Tests
This is to be tested with https://github.com/Expensify/App/pull/17564

# QA
No QA

### CODE REVIEW

**Comment by puneetlath** [comment]
This makes sense to me. Let's add tests for it though.

File: lib/Url.js [hunk 1]
```diff
@@ -1,16 +1,22 @@
...
+const ALLOWED_PORTS = '([1-9][0-9]{0,3}|[1-5][0-9]{4}|6[0-4][0-9]{3}|65[0-4][0-9]{2}|655[0-2][0-9]|6553[0-5])';
+const URL_PROTOCOL_REGEX = '((ht|f|sm)tps?:\\/\\/)';
+const URL_WEBSITE_REGEX = `${URL_PROTOCOL_REGEX}?((?:www\\.)?[-a-z0-9]+?\\.)+(?:${TLD_REGEX})(?:\\:${ALLOWED_PORTS}|\\b|(?=_))`;
 const addEscapedChar = reg => `(?:${reg}|&(?:amp|quot|#x27);)`;
 const URL_PATH_REGEX = `(?:${addEscapedChar('[.,=(+$!*]')}?\\/${addEscapedChar('[-\\w$@.+!*:(),=%~]')}*${addEscapedChar('[-\\w~@:%)]')}|\\/)*`;
 const URL_PARAM_REGEX = `(?:\\?${addEscapedChar('[-\\w$@.+!*()\\/,=%{}:;\\[\\]\\|_]')}*)?`;
 const URL_FRAGMENT_REGEX = `(?:#${addEscapedChar('[-\\w$@.+!*()[\\],=%;\\/:~]')}*)?`;
 const URL_REGEX = `(${URL_WEBSITE_REGEX}${URL_PATH_REGEX}(?:${URL_PARAM_REGEX}|${URL_FRAGMENT_REGEX})*)`;
 
+const URL_REGEX_WITH_REQUIRED_PROTOCOL = URL_REGEX.replace(`${URL_PROTOCOL_REGEX}?`, URL_PROTOCOL_REGEX);
```
Comment (MonilBhavsar) [inline lib/Url.js:12]: We don't consider SMTP a valid protocol in bank account flow

File: __tests__/URL-test.js [hunk 2]
```diff
@@ -0,0 +1,50 @@
+import {URL_REGEX_WITH_REQUIRED_PROTOCOL, URL_REGEX} from '../lib/Url';
+
+
```
Comment (MonilBhavsar) [inline __tests__/URL-test.js:3]: Extra space?

File: __tests__/URL-test.js [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+describe('Mandatory protocol for URL', () => {
+    it('correctly tests valid urls', () => {
+        const regexToTest = new RegExp(`^${URL_REGEX_WITH_REQUIRED_PROTOCOL}$`, 'i');
+        expect(regexToTest.test('https://google.com/')).toBeTruthy();
+        expect(regexToTest.test('http://google.com/')).toBeTruthy();
+        expect(regexToTest.test('ftp://google.com/')).toBeTruthy();
+
```
Comment (MonilBhavsar) [inline __tests__/URL-test.js:10]: I think spaces in between cases can be removed

File: __tests__/URL-test.js [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+        expect(regexToTest.test('https://google.com:65535/path/my')).toBeTruthy();
+    });
+
+    it('correctly tests invalid urls', () => {
+        const regexToTest = new RegExp(`^${URL_REGEX_WITH_REQUIRED_PROTOCOL}$`, 'i');
+        expect(regexToTest.test('google.com')).toBeFalsy();
+
+        expect(regexToTest.test('https://google.com:02')).toBeFalsy();
+        expect(regexToTest.test('https://google.com:65536')).toBeFalsy();
```
Comment (MonilBhavsar) [inline __tests__/URL-test.js:24]: Not sure, if we want to add under this test suite. I think this goes better under invalid URL's?

File: lib/Url.js [hunk 5]
```diff
@@ -1,16 +1,22 @@
 import TLD_REGEX from './tlds';
 
-const URL_WEBSITE_REGEX = `(https?:\\/\\/)?((?:www\\.)?[-a-z0-9]+?\\.)+(?:${TLD_REGEX})(?:\\:\\d{2,4}|\\b|(?=_))`;
+const ALLOWED_PORTS = '([1-9][0-9]{0,3}|[1-5][0-9]{4}|6[0-4][0-9]{3}|65[0-4][0-9]{2}|655[0-2][0-9]|6553[0-5])';
```
Comment (MonilBhavsar) [inline lib/Url.js:3]: Just curious, could you please explain the range of ports covered here?

File: lib/Url.js (hunk 1 above)
Comment (MonilBhavsar) [reply lib/Url.js:12]: Thanks!

Comment (MonilBhavsar) [reply lib/Url.js:12]: Can we also add this to automated test

**Review by puneetlath — DISMISSED** [review]

**Review by MonilBhavsar — COMMENTED** [review]
I just did a quick testing. We should have identified the whole URL including protocol, no?


Screenshot 2023-04-25 at 6 09 50 PM

**Review by puneetlath — APPROVED** [review]
It looks good to me but would like @MonilBhavsar to confirm as well. Thanks!

**Comment by MonilBhavsar** [comment]
Hmm, Looks the lib is linked to project correctly but let me double check.

**Comment by MonilBhavsar** [comment]
Sorry, i think I had issue linking because of conflicting node versions
Looks good and works well. Thanks for working on this!

Screenshot 2023-04-26 at 10 30 17 AM

**Review by MonilBhavsar — APPROVED** [review]

**Pull request merged**